### ISSUE TITLE
Error during APPEND after version update (UNKNOWN-CTE)

### ISSUE THREAD

**Issue by llvilanova** [OfflineIMAP/offlineimap#325]
### General informations

The newer offlineimap seems to choke on many messages that do not exclusively contain plain english text. It seems that things like special symbols on From or even the body switch the message to binary and produce an error (see log).

I found the problem with version 6.7.0+dfsg1-1 after upgrading from 6.3.4-1 (debian package). Downgrading offlineimap makes it work again.

The logs are produced with ``offlineimap -a ac -f test -1 -d all``.

### Log error

```
Account sync ac:
 [imap]:   22:08.09 Account sync ac state => SELECTED
 [imap]:   22:08.09 Account sync ac state_change_pending.release
 [imap]:   22:08.09 Account sync ac _get_untagged_response(EXISTS) => ['0']
 [imap]:   22:08.09 Account sync ac _untagged_response(OK, ?, EXISTS) => ['0']
 [imap]:   22:08.09 Account sync ac [sync] APPEND ('test', '()', '"04-Feb-2016 10:06:56 +0100"')
 [imap]:   22:08.09 Account sync ac state_change_pending.acquire
 [imap]:   22:08.09 Account sync ac _get_untagged_response(OK) => ['[PERMANENTFLAGS (\\Answered \\Flagged \\Draft \\Deleted \\Seen \\*)]  ']
 [imap]:   22:08.09 Account sync ac _get_untagged_response(OK) => ['[UIDVALIDITY 1459693195]  ']
 [imap]:   22:08.09 Account sync ac _get_untagged_response(OK) => ['[UIDNEXT 1]  ']
 [imap]:   22:08.09 Account sync ac _request_push(MHIJ11, APPEND, {}) = MHIJ11
 [imap]:   22:08.09 Account sync ac data=MHIJ11 APPEND test () "04-Feb-2016 10:06:56 +0100" ~{57034}
 [imap]:   22:08.09 Account sync ac _request_push(continuation, APPEND, {}) = MHIJ12
 [imap]:   22:08.09 Account sync ac APPEND:MHIJ12.ready.wait
gw.ac.upc.edu writer:
 [imap]:   22:08.10 gw.ac.upc.edu writer > MHIJ11 APPEND test () "04-Feb-2016 10:06:56 +0100" ~{57034}\r\n
gw.ac.upc.edu reader:
 [imap]:   22:08.10 gw.ac.upc.edu reader poll => [(4, 1)]
 [imap]:   22:08.10 gw.ac.upc.edu reader rcvd 57
 [imap]:   22:08.10 gw.ac.upc.edu reader < MHIJ11 NO [UNKNOWN-CTE] Can not process the binary data\r\n
gw.ac.upc.edu handler:
 [imap]:   22:08.10 gw.ac.upc.edu handler _put_response(MHIJ11 NO [UNKNOWN-CTE] Can not process the binary data)
 [imap]:   22:08.10 gw.ac.upc.edu handler untagged_responses[UNKNOWN-CTE] 0 += [""]
 [imap]:   22:08.10 gw.ac.upc.edu handler _request_pop(MHIJ11, ('NO', ['[UNKNOWN-CTE] Can not process the binary data'])) [1] = MHIJ11
 [imap]:   22:08.10 gw.ac.upc.edu handler APPEND:MHIJ11.ready.set
 [imap]:   22:08.10 gw.ac.upc.edu handler _request_pop(continuation, (False, 'MHIJ11 NO [UNKNOWN-CTE] Can not process the binary data')) [0] = MHIJ12
 [imap]:   22:08.10 gw.ac.upc.edu handler APPEND:MHIJ12.ready.set
 [imap]:   22:08.10 gw.ac.upc.edu handler state_change_free.set
 [imap]:   22:08.10 gw.ac.upc.edu handler NO response: [UNKNOWN-CTE] Can not process the binary data
```

**Comment by nicolas33** [comment]
I think imaplib2 does not support binary mode.

Are all messages failing or only some of them?

**Comment by nicolas33** [comment]
Hmm, v6.3.4 is five years old. imaplib has significant changes since then.

```
c8a511b bump imaplib2 from 2.52 to 2.53
02a5f77 bump imaplib2 from v2.52 to v2.53
04babea bump from imaplib2 v2.51 to v2.52
91d0896 bump imaplib2 from v2.50 to v2.51
2a637bd bump impalib2 from v2.49 to v2.50
cc3cbae bump imaplib2 from v2.48 to v2.49
71dd03e Bump imaplib2 from 2.43 to 2.48
1708fd0 imaplib2: bump to v2.43
ebf1a93 imaplib2: bump to version 2.42
43dbe15 imaplib2: bump to v2.41
7ab779a Imaplib2: use proper tracebacks for exceptions
080317a Imaplib2: trade backticks to repr()
3d88a76 Update imaplib2 to 2.37
be1c72e Updated bundled imaplib2 to 2.36
2df5b71 Updated bundled imaplib2 to 2.35:
d55e4ef imaplib2: fix handling of zero-sized IMAP objects
41cb0f5 Prune trailing whitespaces from code and documentation
2bacdb7 Allow setting IMAP servers' SSL version
8fe8bbe imaplib2: bump upstream version 2.29 --> 2.33
54117e7 Bump bundled imaplib2 to 2.29
1c0c19a imaplib2: bump to upstream version 2.28 (rev 8228a0f)
```

Could you test above commits to track down to the regression?

**Comment by nicolas33** [comment]
Notice that it's weird you to have such issues while others don't. What's your IMAP server domain/software?

**Comment by nicolas33** [comment]
I wonder debian is not using the embedded version of imaplib. imaplib v2.52 is known to have issues in this area.
I wonder offlineimap or imaplib might be patched, too.

@iliastsi can you confirm?

**Comment by nicolas33** [comment]
I meant v2.53 of imaplib is known to have wrong binary support and was removed upstream.

@iliastsi That's what c8a511bb7e48a22f2f1bd is about.
I sucked at the correct commit message, though. Upstream made a new patch to manually come back to the previous v2.52 and I've mixed up release versions while writing the commit message.

This should be:
c8a511b bump imaplib2 from v2.53 to v2.52.

**Comment by nicolas33** [comment]
@llvilanova Could you please test vanilla offlineimap (out of debian package)?
If I'm right, this might be temporary workaround until debian get this fixed.

**Comment by iliastsi** [comment]
@nicolas33 Indeed, offlineimap on Debian is not using the embedded version of imaplib but instead it is using the one provided by Debian. There is an open bug against that package regarding the broken BINARY mode support [1]. I will try to contact the Debian maintainer of imaplib in order to fix that.

@llvilanova You could try and test the vanilla offlineimap, or hot-patch Debian's imaplib2 module
in order to fix the above bug (the patch is here[2]).

[1] https://bugs.debian.org/810245
[2] https://sourceforge.net/p/imaplib2/code/ci/dc98d61fc0013b218590d30f53d10590fcd4256d/

**Comment by llvilanova** [comment]
Sorry, I have a lot of work this week, but will check it on the next one. Thanks!

**Comment by llvilanova** [comment]
@iliastsi Applying the patch seems to fix the problem

Thanks!

**Comment by nicolas33** [comment]
@llvilanova Thanks for the feedback.

@iliastsi Closing since there's a ticket at Debian. Feel free to re-open if you need this one open as well.
